cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -0,0 +1,417 @@
+
+func indexNodeByNodeProviderID(obj interface{}) ([]string, error) {
+	if node, ok := obj.(*apiv1.Node); ok {
+		return []string{node.Spec.ProviderID}, nil
+	}
+	return []string{}, nil
+}
+
+func (c *machineController) findMachine(id string) (*v1alpha1.Machine, error) {
+	item, exists, err := c.machineInformer.Informer().GetStore().GetByKey(id)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, nil
+	}
+
+	machine, ok := item.(*v1alpha1.Machine)
+	if !ok {
+		return nil, fmt.Errorf("internal error; unexpected type %T", machine)
+	}
+
+	return machine.DeepCopy(), nil
+}
+
+func (c *machineController) findMachineDeployment(id string) (*v1alpha1.MachineDeployment, error) {
+	item, exists, err := c.machineDeploymentInformer.Informer().GetStore().GetByKey(id)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, nil
+	}
+
+	machineDeployment, ok := item.(*v1alpha1.MachineDeployment)
+	if !ok {
+		return nil, fmt.Errorf("internal error; unexpected type %T", machineDeployment)
+	}
+
+	return machineDeployment.DeepCopy(), nil
+}
+
+// findMachineOwner returns the machine set owner for machine, or nil
+// if there is no owner. A DeepCopy() of the object is returned on
+// success.
+func (c *machineController) findMachineOwner(machine *v1alpha1.Machine) (*v1alpha1.MachineSet, error) {
+	machineOwnerRef := machineOwnerRef(machine)
+	if machineOwnerRef == nil {
+		return nil, nil
+	}
+
+	store := c.machineSetInformer.Informer().GetStore()
+	item, exists, err := store.GetByKey(fmt.Sprintf("%s/%s", machine.Namespace, machineOwnerRef.Name))
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, nil
+	}
+
+	machineSet, ok := item.(*v1alpha1.MachineSet)
+	if !ok {
+		return nil, fmt.Errorf("internal error; unexpected type: %T", machineSet)
+	}
+
+	if !machineIsOwnedByMachineSet(machine, machineSet) {
+		return nil, nil
+	}
+
+	return machineSet.DeepCopy(), nil
+}
+
+// run starts shared informers and waits for the informer cache to
+// synchronize.
+func (c *machineController) run(stopCh <-chan struct{}) error {
+	c.kubeInformerFactory.Start(stopCh)
+	c.clusterInformerFactory.Start(stopCh)
+
+	klog.V(4).Infof("waiting for caches to sync")
+	if !cache.WaitForCacheSync(stopCh,
+		c.nodeInformer.HasSynced,
+		c.machineInformer.Informer().HasSynced,
+		c.machineSetInformer.Informer().HasSynced,
+		c.machineDeploymentInformer.Informer().HasSynced) {
+		return fmt.Errorf("syncing caches failed")
+	}
+
+	return nil
+}
+
+// findMachineByNodeProviderID find associated machine using
+// node.Spec.ProviderID as the key. Returns nil if either the Node by
+// node.Spec.ProviderID cannot be found or if the node has no machine
+// annotation. A DeepCopy() of the object is returned on success.
+func (c *machineController) findMachineByNodeProviderID(node *apiv1.Node) (*v1alpha1.Machine, error) {
+	objs, err := c.nodeInformer.GetIndexer().ByIndex(nodeProviderIDIndex, node.Spec.ProviderID)
```

Comment:
This is trying to get the underlying machine that is backing the node hence the `findMachine...()` in the func name. 

As the node informer runs we add an additional indexer for all nodes discovered:

```go
func indexNodeByNodeProviderID(obj interface{}) ([]string, error) {
	if node, ok := obj.(*apiv1.Node); ok {
		return []string{node.Spec.ProviderID}, nil
	}
	return []string{}, nil
}
```

This function uses the `node.Spec.ProviderID` which is they key we are using in this function. This gives us O(1) lookup time when mapping between a node and its underlying machine.

Reply:
> It looks to me like you are trying to get the node you already have or am I misunderstanding it?

Although I replied, I think I need to take a look at this again.

Reply:
@alvaroaleman 
The node here is an artificially created node, which has just providerID. Informer here trying to get actual node object using this providerID, though i think this should not be done. Please refer to the discussion here, https://github.com/openshift/kubernetes-autoscaler/pull/93#discussion_r285180838

Reply:
I did the change[1] in the OpenShift version first - we can cherry-pick the changes from there in due course when it is has been validated beyond just the unit tests. 

[1] - https://github.com/openshift/kubernetes-autoscaler/pull/97

Reply:
> The node here is an artificially created node, which has just providerID. Informer here trying to get actual node object using this providerID, though i think this should not be done. Please refer to the discussion here,

That is wrong, the node does have the Annotation. The current implementation does not work when the node has no providerID.  Removing the first part of the func like this fixes that:

```
 func (c *machineController) findMachineByNodeProviderID(node *corev1.Node) (*v1alpha1.Machine, error) {
-       objs, err := c.nodeInformer.GetIndexer().ByIndex(nodeProviderIDIndex, node.Spec.ProviderID)
-       if err != nil {
-               return nil, err
-       }
-
-       switch n := len(objs); {
-       case n == 0:
-               return nil, nil
-       case n > 1:
-               return nil, fmt.Errorf("internal error; expected len==1, got %v", n)
-       }
-
-       node, ok := objs[0].(*corev1.Node)
-       if !ok {
-               return nil, fmt.Errorf("internal error; unexpected type %T", node)
-       }
-
        if machineName, found := node.Annotations[machineAnnotationKey]; found {
                return c.findMachine(machineName)
        }
```

Reply:
This has been reworked here: https://github.com/kubernetes/autoscaler/pull/1866/commits/8b32581909d286afab1b638ac3d17a1043bfcca2#diff-89363471d762322b93c9bcfec57b5af8R160

Reply:
@alvaroaleman I think via a slack discussion we agreed this is OK now. Requires ProviderID to be set because that's the expectation for identifying nodes from the CA's perspective, particularly when determining if the cluster is in a healthy state.